### PR TITLE
notification: implement ClientV1 for receiving notifications

### PR DESCRIPTION
**Pull request by unknwon**
Part of CORE-92

This PR implements a new notification (v1) subscriber for receiving SAMS notifications. It is pretty simple since the heavy work is done by the pubsub SDK, and we currently only have one single event ("UserDeleted").

> [!note]
> ~~If I get managed to merge https://github.com/sourcegraph/sourcegraph/pull/62136 before this, I will also make the `Client` to return a `ReceiveRoutine` for better DX. Otherwise, I will do this as a follow-up.~~

## Test plan

Integrated into SAMS by sending notifications to a test Slack channel (also serve as the dogfood):

CleanShot 2024-05-04 at 10 48 31@2x

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -202,6 +202,53 @@ func main() {
...
+	go func() {
+		err := client.Receive(ctx, handler)
+		if err != nil {
+			logger.Error("failed to receive notification", log.Error(err))
+		}
+	}()
+
+	// For demonstration purposes, we will run the client for 1 minute.
+	time.Sleep(time.Minute)
+	cancel() // Cancel the context to stop receiving notifications.
```
Comment (bobheadxi) [inline README.md:248]: I think the more common (and easier-to-hold, IMO) pattern is a `.Close() error` on the `client`, instead of context cancellation

File: notifications/v1/types.go [hunk 2]
```diff
@@ -0,0 +1,16 @@
+package v1
+
+// ⚠️ WARNING: These types MUST match the SAMS implementation, at
+// backend/internal/notification/types.go
+
+const (
+	nameUserDeleted = "UserDeleted"
+)
```
Comment (bobheadxi) [inline notifications/v1/types.go:8]: Should these be exported, and then the SAMS implementation can reference them when publishing events?

File: README.md (hunk 1 above)
Comment (unknwon) [reply README.md:248]: Good idea :D

File: notifications/v1/types.go (hunk 2 above)
Comment (unknwon) [reply notifications/v1/types.go:8]: I actually prefer try not to do that, it sounds very reasonable thing to do but in my experience, it's a nasty dependency to deal with (i.e. have to make a PR on this repo just for a string const, then go mod replace on the SAMS repo with a broken PR. This could totally be Go's toolchain fault, but I don't feel strong that it's worth it doing it now)

File: README.md (hunk 1 above)
Comment (unknwon) [reply README.md]: Added in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/3cabd201762491d905828bea83932f42599c9987!

File: notifications/v1/types.go (hunk 2 above)
Comment (bobheadxi) [reply notifications/v1/types.go:8]: I mean string const addition is basically an API addition, i.e. "this is coming soon", and is detached from implementation, like any proposed change to the API spec. WDYT?

File: notifications/v1/client.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+// Receive starts receiving SAMS notifications and calls the corresponding
+// handler for each notification. It blocks until the context is done (e.g.
+// deadline exceed or canceled).
+//
+// ⚠️ WARNING: Each subscription can only have one active receiver at a time,
+// i.e. there should only be one call to Receive for a given subscription.
+func (c *Client) Receive(handler *ReceiveHandlers) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancelContext = cancel
+	return c.subscription.Receive(ctx, func(ctx context.Context, msg *pubsub.Message) {
```
Comment (bobheadxi) [inline notifications/v1/client.go:53]: Can we receive messages concurrently, or is it only ever 1 at a time? Do we want to support concurrent handling, or is that a handler implementation details? Should handlers expect to be concurrency-safe?

File: notifications/v1/client.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
+	OnUserDeleted func(data *UserDeletedData) error
+}
+
+// Receive starts receiving SAMS notifications and calls the corresponding
+// handler for each notification. It blocks until the context is done (e.g.
+// deadline exceed or canceled).
+//
+// ⚠️ WARNING: Each subscription can only have one active receiver at a time,
+// i.e. there should only be one call to Receive for a given subscription.
```
Comment (bobheadxi) [inline notifications/v1/client.go:49]: I think this should panic if a handler has already been registered then

But, I also don't see a reason _not_ to support multiple receivers - we can have `[]handler` and iterate over those when receiving messages.

In that case you might want an API surface like:

```go
RegisterReceiver(handler)
Start() error
```

If you want to keep single-receiver-only, IMO you should add it to `NewClient`, then:

```
NewClient(...)
(c) Start() error
(c) Stop() error
```

File: notifications/v1/client.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+// ReceiveHandlers is a collection of receive handlers for each type of SAMS
+// notifications. If the handler of a notification is nil, the notification will
+// be acknowledged automatically without any processing.
+//
+// If a handler returns an error, the notification will be unacknowledged and
+// retried later.
+type ReceiveHandlers struct {
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
+	OnUserDeleted func(data *UserDeletedData) error
```
Comment (bobheadxi) [inline notifications/v1/client.go:41]: WDYT about just `Handler` as a name, for `notifications.Handlers.OnUserDeleted`?

File: notifications/v1/client.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+			return
+		}
+
+		err = c.handleReceive(handler, msgData.Name, msgData.Metadata)
+		if err != nil {
+			c.logger.Error("failed to process notification message", log.Error(err))
+			msg.Nack()
+			return
+		}
+		msg.Ack()
```
Comment (chrsmith) [inline notifications/v1/client.go:71]: What do you think about the following? IMHO it's a little clearer, since both "branches" if the `if` statement need to deal with responding to the message in some way. But having both `if` branches makes this super-clear.


```suggestion
		err = c.handleReceive(handler, msgData.Name, msgData.Metadata)
		if err == nil {
			msg.Ack()
		} else {
			c.logger.Error("failed to process notification message", log.Error(err))
			msg.Nack()
		}
```

File: notifications/v1/client.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	})
+}
+
+func (c *Client) handleReceive(handler *ReceiveHandlers, name string, metadata json.RawMessage) error {
+	switch name {
+	case nameUserDeleted:
+		var data UserDeletedData
+		err := json.Unmarshal(metadata, &data)
+		if err != nil {
```
Comment (chrsmith) [inline notifications/v1/client.go:80]: Super-duper-minor suggestion: The code you have is fine as-is, but since line 79 is so short, it is (arguably, marginally) cleaner to not introduce `err` into the higher scope.

```suggestion
		if err := json.Unmarshal(metadata, &data); err != nil {
```

File: notifications/v1/client.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+			return errors.Wrap(err, "unmarshal metadata")
+		}
+		if handler.OnUserDeleted != nil {
+			return handler.OnUserDeleted(&data)
```
Comment (chrsmith) [inline notifications/v1/client.go:84]: Any reason why we don't move the `json.Unmarshal` code to below the `handler.OnUserDeleted != nil` line? e.g. suppose the data is corrupted in some way... there isn't any reason to have that surface an error if the handler isn't even listening to that event anyways. Right?

File: notifications/v1/types.go (hunk 2 above)
Comment (chrsmith) [reply notifications/v1/types.go:8]: (Apologies in advance if I'm totally missing something here.)

I'd be inclined to avoid the circular dependency between the SAMS SDK and SAMS itself, as it does seem like it would complicate things for little-to-know benefit.

Ultimately consumers of the SAMS SDK _shouldn't_ know anything about the topic names, because it's irrelevant to their needs. As someone hooking into this, I just need to be concerned about using the `type ReceiveHandlers struct`. And registering the appropriate `OnEventX` handlers. The actual names used in the Pub/Sub topics is 100% irrelevant.

> I mean string const addition is basically an API addition, i.e. "this is coming soon", and is detached from implementation, like any proposed change to the API spec. WDYT?

If SAMS starts or stops submitting events with a new name, clients don't really care. It should be a warning-level log, since it indicates that the client is running on an older build of the SAMS SDK, but there being events that are in-the works doesn't really mean much.

i.e. as a consumer of this library, either I provide an `OnEventX` handler to start processing those events, or not. The current state of rollout on the SAMS-end isn't relevant?

File: notifications/v1/types.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+// UserDeletedData contains information of a "UserDeleted" notification.
+type UserDeletedData struct {
+	// UserID is the SAMS external user ID of the deleted user.
+	UserID string `json:"user_id"`
```
Comment (chrsmith) [inline notifications/v1/types.go:13]: `user_id` isn't really an idiomatic name for a JSON field. Is there a particular reason we are using that instead of `userId`? Similarly, does this actually matter as long as SAMS and the SAMS SDK are in agreement (e.g. it could be `xxxasdfkjl2yolo`)?

File: notifications/v1/types.go [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+// ⚠️ WARNING: These types MUST match the SAMS implementation, at
+// backend/internal/notification/types.go
+
+const (
+	nameUserDeleted = "UserDeleted"
+)
+
+// UserDeletedData contains information of a "UserDeleted" notification.
+type UserDeletedData struct {
+	// UserID is the SAMS external user ID of the deleted user.
```
Comment (chrsmith) [inline notifications/v1/types.go:12]: #namingishard

In working on SSC bits and integrating SAMS with other services, I've referred to this identifier a number of ways:

- SAMS external ID
- SAMS account external ID
- SAMS ID
- Account ID

... but I've personally never used "User ID" / "SAMS User ID". That isn't a good reason to suggest renaming this. But as an alternative, what do you think about `AccountID`. As in:

```go
// AccountID is the SAMS external ID for the deleted user's SAMS account.
```

**Review by chrsmith — APPROVED** [review]
Lots of minor feedback, but nothing major.

The only "problem" I see is just how I track down the `projectId/subscriptionId` / `GCP_PROJECT`/`SAMS_NOTIFICATION_SUBSCRIPTION` values.

🤞 that these can be set automatically by the MSP machinery, provided I set something like `sams_environment = "dev"` or something.

File: notifications/v1/client.go (hunk 4 above)
Comment (bobheadxi) [reply notifications/v1/client.go:49]: > So I'd (gently, humbly, kindly) suggest that we keep the SAMS SDK as simple as possible, to make it easier for clients to use safely/correctly. (And at least in this case, it doesn't seem like we are preventing more exotic/advanced use cases.)

Agree - I think adding `handler` as something you must provide up-front when creating a client is probably the best way to go, i.e. the API surface would be:

```go
NewClient(...) // provide handler here
(c) Start() error // long-lived routine subscribed to the events
(c) Stop() error // shutdown
```

Maybe `client` can just implement the background worker? Then it can start and stop with the MSP service runtime as part of the collection of routines :)

File: notifications/v1/types.go (hunk 2 above)
Comment (bobheadxi) [reply notifications/v1/types.go:8]: Fair enough, but `⚠️ WARNING: These types MUST match the SAMS implementation, at X`-type comments feels like an uncomfortable invariant for me, irrespective of "does it matter or not".

Also, for "clients" in terms of "user calling the SDK" indeed they shouldn't care, but for "clients" in terms of "the implementation of the SDK being a client to SAMS", it certainly does care about what gets published, which is why this is here in the first place

File: notifications/v1/client.go [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+// subscription.
+type Client struct {
+	logger        log.Logger
+	subscription  *pubsub.Subscription
+	cancelContext context.CancelFunc
+}
+
+// NewClient creates a new Client for receiving SAMS notifications with given
+// GCP project ID and Pub/Sub subscription ID.
+func NewClient(logger log.Logger, projectID, subscriptionID string) (*Client, error) {
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply notifications/v1/client.go:22]: @unknwon we should offer standardized constructor-from-env setup like #15 for this

File: notifications/v1/client.go (hunk 4 above)
Comment (chrsmith) [reply notifications/v1/client.go:49]: > Maybe client can just implement the background job interface? Then it can start and stop with the MSP service runtime as part of the collection of routines :)

... that actually would be huge. 😍 

In the limit, every service eventually grows to the point that it needs to have some sort of "background job" / "message queue" system. And if there is some ultra-lightweight version that can be provided by the MSP runtime, that would be a capability that would be useful to MSP services in the future.

e.g. it is non-trivial to correctly/safely do the wiring to gracefully stop something. (e.g. when to call `Stop() error`, and make sure we call it in any number of situations -- e.g. SIGTERM, or some GCP-specific event, etc.) So being able to just delegate that to `runtime.RegisterShutdownCallback(logger, pubSubClient.Stop)` would mean that it's one less thing that we need to implement N-different times across various services.

Comment (bobheadxi) [reply notifications/v1/client.go:49]: This already exists, you just have to return a multi-routine here https://sourcegraph.sourcegraph.com/github.com/sourcegraph/self-serve-cody/-/blob/cmd/accounts-server/service.go?L49-65, e.g. https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/telemetry-gateway/service/service.go?L119-138 - the runtime spins up the routine and does intercept handling and shutdown logic via `lib/background`, which runs all the single-tenant Sourcegraph services today

File: notifications/v1/types.go (hunk 2 above)
Comment (unknwon) [reply notifications/v1/types.go:8]: > Fair enough, but `⚠️ WARNING: These types MUST match the SAMS implementation, at X`-type comments feels like an uncomfortable invariant for me, irrespective of "does it matter to clients or not".

We can reword this and change it to a note instead to something like "see xxxx for the full list of supported notifications". Does that make you more comfortable? 😁

File: notifications/v1/client.go (hunk 3 above)
Comment (unknwon) [reply notifications/v1/client.go:53]: Good call on OOTB concurrency control (currently 1 at a time)! But I think whether or not handlers can be concurrency-safe is defined by each handler implementation (i.e. do they do it in a concurrency-safe manner).

Comment (unknwon) [reply notifications/v1/client.go:53]: Actually, re-read the `subscription.Receive` docstring, it says

>Receive calls f concurrently from multiple goroutines.

but it does provide an option to sort of limiting the max concurrency.

Comment (unknwon) [reply notifications/v1/client.go:59]: I decided to expose the `pubsub.ReceiveSettings` in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/08bb5832ea6edd81e434500f7b49d4b9dabcce40, PTAL!

File: notifications/v1/client.go (hunk 4 above)
Comment (unknwon) [reply notifications/v1/client.go]: Sorry for the confusion here 😅 I updated the docstring to explain that this constraint is imposed by `pubsub.Subscription.Receive` 😂 not me... as Chris said, if we were able to do it, it only makes the use case easier. For now, people can still create multiple `Client`s (that comes with different `pubsub.Subscription`s) to achieve the same effect if they _really_ want to.

---

On the topic of "Maybe client can just implement the background job interface", that's actually my plan, once https://github.com/sourcegraph/sourcegraph/pull/62136 is able to get merged (which may take another 2-3 weeks...)

File: notifications/v1/client.go (hunk 5 above)
Comment (unknwon) [reply notifications/v1/client.go:41]: I started with `Handler`, but I prefer more concrete naming given this is after all, an argument being passed to a method (whose name is "Receive"). Also, in the future, I think it's 50/50 chance we might support Push subscription, then the naming of "Handler" might create ambiguity. WDYT?

**Comment by unknwon** [comment]
> The only "problem" I see is just how I track down the `projectId/subscriptionId` / `GCP_PROJECT`/`SAMS_NOTIFICATION_SUBSCRIPTION` values.
> 
> 🤞 that these can be set automatically by the MSP machinery, provided I set something like `sams_environment = "dev"` or something.

Totally understand this yet to be obvious (until I flesh out the integration docs in our handbook), because we use Pull subscription here, each subscriber is expected to create their own Pub/Sub subscriptions in their own GCP projects, the TF would look almost identical to https://github.com/sourcegraph/managed-services/blob/46de5cf3b86e54aa4d31181f08e00f8e04eb8beb/services/sourcegraph-accounts/terraform/prod/stacks/cloudrun/pubsub.tf#L16-L44 (the SAMS own dogfood subscription, the only change would be the `topic`) because of the 1-N relation. Otherwise, only one subscriber would get the notification if they share the same subscription.

Therefore you can tell now it is pretty much freeform. Although in MSP, we can push for more conventional setup (if we found it to be suitable) 😁

File: notifications/v1/client.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (unknwon) [reply notifications/v1/client.go:22]: Please see my comment in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24#issuecomment-2100712267.

>Or are these things I would need to wire into my MSP service configuration?

TLDR yes, I think things would be clearer to be documented in the upcoming integration docs than docstring here (because here it missing the context of provisioning the actual Pub/Sub subscription).

LMK what do you think, can be convinced 😁

[the remark this answers is not in the document]
Comment (unknwon) [reply notifications/v1/client.go:22]: > maybe we should offer standardized constructor-from-env setup like #15 for this?

👍 good idea, will push a commit.

File: notifications/v1/client.go (hunk 6 above)
Comment (unknwon) [reply notifications/v1/client.go:80]: I am a weakly-held fan of early returns, but 🤷‍♂️ 😁

File: notifications/v1/client.go (hunk 7 above)
Comment (unknwon) [reply notifications/v1/client.go:89]: Sure, if it can make you happier!

File: notifications/v1/client.go (hunk 8 above)
Comment (unknwon) [reply notifications/v1/client.go]: Good catch! Updated in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/95a8735aa27709ba71466e99f2bed134cd0f9e81

File: notifications/v1/types.go (hunk 9 above)
Comment (unknwon) [reply notifications/v1/types.go:13]: > `user_id` isn't really an idiomatic name for a JSON field.

Really? 😂 For all my life I have believed that's the convention (maybe I used too much GitHub API?), and all the OAuth/OIDC fields uses snake case 🤔

Comment (chrsmith) [reply notifications/v1/types.go:13]: OK, perhaps you are right and I'm speaking nonsense. (Sorry.) Apparently my suggestion is only 0.33333 reasonable 😭 .

1. If the RFC or some spec written by people much smarter than me does it some way, we should do it that way too. (So ❌ for Chris.)
2. Anything GitHub does is perfect. Therefore, if that is how they name identifiers in their API, it is how things should be done. (Though I'd guess it has something to do with the service's roots in Ruby...) (So ❌ for Chris.)
3. JSON objects are used at runtime in JavaScript, and therefore would be subject to the Google JavaScript Style Guide. (So 🟢 for Chris.)

https://google.github.io/styleguide/jsguide.html#naming-non-constant-field-names

Comment (bobheadxi) [reply notifications/v1/types.go:13]: I always associated fooBar with JSON and foo_bar with YAML, to throw some noise into the debate :)

File: notifications/v1/types.go (hunk 2 above)
Comment (bobheadxi) [reply notifications/v1/types.go:8]: I mean the implication is still a "we should match some hand-maintained invariant", but sure, not a huge deal

File: notifications/v1/client.go (hunk 4 above)
Comment (bobheadxi) [reply notifications/v1/client.go]: > But I do not want to force people to use background.Routine, so in my plan, the Client would return a ReceiveRoutine that implements the background.Routine they can grab and go.

Can we do this now, and just update this implementation later once https://github.com/sourcegraph/sourcegraph/pull/62136 lands? The current implementation IMO is hard to hold right - I think we should just have a new `ReceiverClient` that is a background routine, and users don't have to worry about starting and stopping it. Starting and stopping the receiver shouldn't be an application concern, it should match 1:1 with service lifecycle, is there ever a situation where it shouldn't?

File: notifications/v1/client.go (hunk 3 above)
Comment (bobheadxi) [reply notifications/v1/client.go:62]: Depends on outcome of https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24#discussion_r1594295009

Comment (bobheadxi) [reply notifications/v1/client.go:62]: LGTM, though we may want to make some changes depending on discussion in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24#discussion_r1594295009

**Comment by chrsmith** [comment]
> Totally understand this yet to be obvious (until I flesh out the integration docs in our handbook), because we use Pull subscription here, each subscriber is expected to create their own Pub/Sub subscriptions in their own GCP projects, the TF would look almost identical to https://github.com/sourcegraph/managed-services/blob/46de5cf3b86e54aa4d31181f08e00f8e04eb8beb/services/sourcegraph-accounts/terraform/prod/stacks/cloudrun/pubsub.tf#L16-L44 

GOSH! That sucks.

Why can't we write a single, reusable software module so we can make it easier to create this specific kind of cloud infrastructure? Terraform may technically be able to do that... but HCL isn't a viable language. If only we could use a familiar, programming language for writing this type of infrastructure.

Heck! If we were to write some, e.g., TypeScript code for creating a `SAMSPubSubSubscription` resource, we could even write unit-tests to ensure that the resource was created with the correct settings...

I've heard that there is this HOT NEW STARTUP called [Pulumi](www.pulumi.com). Perhaps their product might be able to solve this problem, in a robust and scalable way?

File: notifications/v1/types.go (hunk 9 above)
Comment (unknwon) [reply notifications/v1/types.go:13]: I do notice convention is JS world and non-JS is very different, my understanding is that's simply because by using camel case in JS, that JSON field can be referenced directly as a field of an object (`T.userId` ✅ ), but not so if using snake case (`T.user_id` ❌ ;GraphQL counts as JS world too).

However, outside JS world, I've only found snake case to have conventional look and feel.

File: notifications/v1/types.go (hunk 10 above)
Comment (unknwon) [reply notifications/v1/types.go:12]: Fair! Gonna pick "AccountID" 😁

File: README.md [hunk 12]
```diff
@@ -210,19 +210,26 @@ The SAMS Notifications API is for distributing notifications to downstream syste
...
 	notificationv1 "github.com/sourcegraph/sourcegraph-accounts-sdk-go/notifications/v1"
 )
 
 func main() {
 	var logger log.Logger // TODO: Initialize your logger.
 
-	client, err := notificationv1.NewClient(logger, os.Getenv("GCP_PROJECT"), os.Getenv("SAMS_NOTIFICATION_SUBSCRIPTION"))
+	client, err := sams.NewNotificationsV1(
+		logger,
+		// In MSP, you can use `sams.NewNotificationsV1ConfigFromEnv` to get a `sams.NotificationsV1Config`.
```
Comment (chrsmith) [inline README.md]: 👍 this is super-helpful!

File: notifications/v1/client.go [hunk 13]
```diff
@@ -6,90 +6,153 @@ import (
...
-type Client struct {
-	logger        log.Logger
-	subscription  *pubsub.Subscription
+type subscriber struct {
+	logger       log.Logger
+	handlers     SubscriberHandlers
+	subscription *pubsub.Subscription
+
+	// state indicates the state of workers, must be one of "idle", "started" or
+	// "stopped".
```
Comment (chrsmith) [inline notifications/v1/client.go:20]: No changes needed, but something to consider:

These 3x states to form a basic state machine? e.g. Only a subscriber in the "idle" state can call `Start()`,  and only in the "started" state can you call `Stop()`.

The doc comment here is good, since it calls out something important that isn't clear from the code. But it may perhaps be useful to add another couple of sentences for what those states _mean_, and what invariants we expect to maintain.

Given the simplicity/clarify of this, updating the comment wouldn't be adding any real value. But you can see if there was a fourth state, that some other type of abstraction or clarification about that underlying state machine would be important.

File: notifications/v1/client.go [hunk 14]
```diff
@@ -6,90 +6,153 @@ import (
...
+	// zero value means to use the default settings.
+	ReceiveSettings ReceiveSettings
+	// Handlers is the collection of subscription handlers for each type of SAMS
+	// notifications.
+	Handlers SubscriberHandlers
+}
+
+func (opts SubscriberOptions) Validate() error {
+	if opts.ProjectID == "" {
+		return errors.New("ProjectID is required")
```
Comment (chrsmith) [inline notifications/v1/client.go:45]: Kinda surprised the linter isn't complaining that the error message starts with a capital letter. (i.e. this error is "good", but a situation where linters usually flag it as a false positive... so usually I find that I need to write something lame like "**property** X is required".)

But maybe linters are smart enough now?

File: notifications/v1/client.go [hunk 15]
```diff
@@ -6,90 +6,153 @@ import (
...
+	}
+}
+
+func (r *subscriber) Stop() {
+	if !r.state.CompareAndSwap(stateStarted, stateStopped) {
+		r.logger.Error("failed to stop subscriber", log.Error(errors.New(`not in "started" state`)))
+		return
+	}
+
+	r.cancelContext()
```
Comment (chrsmith) [inline notifications/v1/client.go:117]: Should we have a `nil` check here in case `cancelContext` wasn't set? For example, on line 69 a `subscriber` is created _without_ specifying the cancel context. So if we are expecting this value to "come later", then just to err on the side of caution, then perhaps we should anticipate it not being set at all?

Though, thinking about it, I guess asserting that we are in the "stopped" state may be sufficient to ensure that `Start()` was called... and perhaps that is where `cancelContext` would be set?

File: notifications/v1/client.go [hunk 16]
```diff
@@ -6,90 +6,153 @@ import (
...
 
-func (c *Client) handleReceive(handler *ReceiveHandlers, name string, metadata json.RawMessage) error {
+// SubscriberHandlers is a collection of subscription handlers for each type of
+// SAMS notifications. If the handler of a notification is nil, the notification
+// will be acknowledged automatically without any processing.
+//
+// If a handler returns an error, the notification will be unacknowledged and
+// retried later.
+type SubscriberHandlers struct {
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
```
Comment (chrsmith) [inline notifications/v1/client.go:128]: Where is the "most canonical" documentation for this that you expect people who integrate with this code?

If we expect that people who are wiring up a `SubscriptionHandler` for a SAMS-dependent service, I would suggest we try to expand on the doc comments here to help explain when/why that "UserDeleted" notification would be received.

... not that it is confusing in this particular situation, but I'd nudge you to do that in this first commit so that we will follow that pattern as we add new events. Because when the event is more ambiguous, e.g. `OnUserProfileUpdated` or `OnAccountVerified`, putting a 1-2 sentences explanation would make it easier for people to integrate with this correctly. (e.g. understanding what to expect and not-expect when they see this event.)

File: notifications/v1/client.go [hunk 17]
```diff
@@ -6,90 +6,153 @@ import (
...
+		r.logger.Warn("acknowledging unknown notification name", log.String("name", name))
 	}
 	return nil
 }
 
-// Close stops the client from receiving notifications.
-func (c *Client) Close() {
-	c.cancelContext()
-}
+const (
```
Comment (chrsmith) [inline notifications/v1/client.go:154]: I, being brain damaged by the functional programming disease, an morally obligated to point out that we can TOTALLY USE A TYPE HERE.

```go
type subscriberState string
const (
	stateIdle    = "idle"
	stateStarted = "started"
	stateStopped = "stopped"
)
```

FAQ answers:

1. Yes, this isn't actually of much benefit.
2. Yes, this will require you to add `string(...)` casts in various places.
3. Yes, as a card carrying functional programming wonk, I will have my membership status revoked unless I point out any and every opportunity to lean on the type checker to catch things like this.
4. No, I don't care if you ignore this suggestion. In fact, I would be annoyed if someone asked me to. But see point number 3.

File: notifications/v1/types.go [hunk 18]
```diff
@@ -9,8 +9,8 @@ const (
 
 // UserDeletedData contains information of a "UserDeleted" notification.
 type UserDeletedData struct {
-	// UserID is the SAMS external user ID of the deleted user.
-	UserID string `json:"user_id"`
+	// AccountID is the SAMS external ID of the deleted user.
+	AccountID string `json:"account_id"`
```
Comment (chrsmith) [inline notifications/v1/types.go:13]: 🙏

**Review by chrsmith — APPROVED** [review]
Took a quick Look at the recent changes. Other than one potential panic if there happens to be a codepath where a field isn't set, everything LGTM!

Thank you for applying earlier suggestions, IMHO things are a bit easier to follow now.

File: notifications/v1/client.go [hunk 19]
```diff
@@ -0,0 +1,158 @@
...
+	})
+	if err != nil {
+		s.logger.Error("failed to receive notifications", log.Error(err))
+		return
+	}
+}
+
+func (r *subscriber) Stop() {
+	if !r.state.CompareAndSwap(stateStarted, stateStopped) {
+		r.logger.Error("failed to stop subscriber", log.Error(errors.New(`not in "started" state`)))
```
Comment (bobheadxi) [inline notifications/v1/client.go:113]: same here, IMO this should just be a panic

File: notifications/v1/client.go [hunk 20]
```diff
@@ -0,0 +1,158 @@
...
+		logger:       logger.Scoped("notification.subscriber"),
+		handlers:     opts.Handlers,
+		subscription: subscription,
+		state:        atomic.NewString(stateIdle),
+	}, nil
+}
+
+func (s *subscriber) Start() {
+	if !s.state.CompareAndSwap(stateIdle, stateStarted) {
+		s.logger.Error("failed to start subscriber", log.Error(errors.New(`not in "idle" state`)))
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply notifications/v1/client.go:79]: or, maybe we can let this be idempotent, and just return with a no-op

File: notifications/v1/client.go (hunk 19 above)
Comment (bobheadxi) [reply notifications/v1/client.go:113]: or https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/files#r1595711888

File: notifications/v1/client.go (hunk 15 above)
Comment (bobheadxi) [reply notifications/v1/client.go:117]: I think the state check is sufficient, though it may help to have a docstring here indicating as such

**Review by bobheadxi — APPROVED** [review]

**Comment by bobheadxi** [comment]
If you want, we should also provide a MSP runtime example where you just return this as part of service start (which expects a routine, which can be a CombinedRoutine or LIFOStopRoutine)

File: notifications/v1/client.go (hunk 13 above)
Comment (unknwon) [reply notifications/v1/client.go]: Actually pushed a commit to make such simple state machine https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/cd459a3a207d91257f2567979583f82de3f593f3 😂 (I failed the computation class that teaches the state machine so... LMK WDYT)

File: notifications/v1/client.go (hunk 14 above)
Comment (unknwon) [reply notifications/v1/client.go]: Haha yeah, sometimes the linear is smart enough (assuming we did enable the one you're referring to 😂

File: notifications/v1/client.go (hunk 15 above)
Comment (unknwon) [reply notifications/v1/client.go]: Good catch, tho I'd generally say state check is enough here, but I pushed a commit https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/75628f4861c26807eae94b5f4e37347d6f9fc0c8 that always gives `cancelContext` a default complaining it is called without being set (with a useful value) as the simple safeguard.

File: notifications/v1/client.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (chrsmith) [reply notifications/v1/client.go]: +1 to panicing. At this point, there is no way for the system to recover. So just "halt and catch fire" is for the best. (Since otherwise, it's possible for this to maybe make it through unit tests, CI/CD, and explode later than we'd like.)

File: notifications/v1/client.go (hunk 16 above)
Comment (unknwon) [reply notifications/v1/client.go]: Enriched the docstring in https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/24/commits/e1a5374399524832b1f3444c034e3b40c1687d00!

File: notifications/v1/client.go (hunk 20 above)
[the remark this answers is not in the document]
Comment (unknwon) [reply notifications/v1/client.go]: The way it is doing is practically noop (doesn't try to fail anything) but I think also surfacing the error would be enough for the system owner notice something is wrong that they need to fix.

[the remark this answers is not in the document]
Comment (unknwon) [reply notifications/v1/client.go]: Alright, since I got two votes for panic, let's panic :D

**Comment by unknwon** [comment]
> If you want, we should also provide a MSP runtime example where you just return this as part of service start (which expects a routine, which can be a CombinedRoutine or LIFOStopRoutine)

Let me take a note and look into that as a follow-up.

**Comment by unknwon** [comment]
Thanks for all the reviews! Merging as-is, happy to address post-merge feedback!

**Comment by unknwon** [comment]
> > If you want, we should also provide a MSP runtime example where you just return this as part of service start (which expects a routine, which can be a CombinedRoutine or LIFOStopRoutine)
> 
> Let me take a note and look into that as a follow-up.

Because the handbook directly included MSP examples, so I decided to add a link pointing to that instead, see https://github.com/sourcegraph/sourcegraph-accounts-sdk-go/pull/31
